luigi/rpc.py

```diff
@@ -144,35 +144,29 @@ def __init__(self, url='http://localhost:8082/', connect_timeout=None):
         else:
             self._fetcher = URLLibFetcher()
 
-    def _wait(self):
-        if self._rpc_log_retries:
-            logger.info("Wait for %d seconds" % self._rpc_retry_wait)
-        time.sleep(self._rpc_retry_wait)
+    def _get_retryer(self):
+        def retry_logging(retry_state):
+            if self._rpc_log_retries:
+                logger.warning("Failed connecting to remote scheduler %r", self._url, exc_info=True)
+                logger.info("Retrying attempt %r of %r (max)" % (retry_state.attempt_number + 1, self._rpc_retry_attempts))
+                logger.info("Wait for %d seconds" % self._rpc_retry_wait)
+
+        return Retrying(wait=wait_fixed(self._rpc_retry_wait),
+                        stop=stop_after_attempt(self._rpc_retry_attempts),
+                        reraise=True,
+                        after=retry_logging)
 
     def _fetch(self, url_suffix, body):
         full_url = _urljoin(self._url, url_suffix)
-        last_exception = None
-        attempt = 0
-        while attempt < self._rpc_retry_attempts:
-            attempt += 1
-            if last_exception:
-                if self._rpc_log_retries:
-                    logger.info("Retrying attempt %r of %r (max)" % (attempt, self._rpc_retry_attempts))
-                self._wait()  # wait for a bit and retry
-            try:
-                response = self._fetcher.fetch(full_url, body, self._connect_timeout)
-                break
-            except self._fetcher.raises as e:
-                last_exception = e
-                if self._rpc_log_retries:
-                    logger.warning("Failed connecting to remote scheduler %r", self._url,
-                                   exc_info=True)
-                continue
-        else:
+        scheduler_retry = self._get_retryer()
+
+        try:
+            response = scheduler_retry(self._fetcher.fetch, full_url, body, self._connect_timeout)
```

Comment:
for some reason i'm struggling to see where it's defined that `full_uri`, `body`, and `self._connect_timeout` are being passed to `self._fetcher.fetch()` here.

Reply:
Thank you for your comment!
Here is a definition of `Retrying`, first argument of this class' `__call__` is function, and followings are its arguments.
https://github.com/jd/tenacity/blob/3e2244535ccfbb6a4b7fdd77bfc9aa61a1302302/tenacity/__init__.py#L422-L442

Reply:
Ah, i see. A bit obfuscated, but it is being used as tenacity expects